crates/ruma-events/src/secret/request.rs

```diff
@@ -0,0 +1,302 @@
+use ruma_identifiers::DeviceIdBox;
+use ruma_serde::StringEnum;
+use serde::{ser::SerializeStruct, Deserialize, Serialize};
+
+use crate::ToDeviceEvent;
+
+/// Event sent by a client to request a secret from another device or to cancel a previous request.
+///
+/// It is sent as an unencrypted to-device event.
+pub type RequestToDeviceEvent = ToDeviceEvent<RequestToDeviceEventContent>;
+
+/// The payload for RequestToDeviceEvent.
+#[derive(Clone, Debug, Serialize, Deserialize, EventContent)]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+#[ruma_event(type = "m.secret.request", kind = ToDevice)]
+pub struct RequestToDeviceEventContent {
+    /// The action for the request, one of `["request", "request_cancellation"]`.
+    ///
+    /// If the action is "request", the name of the secret must also be provided.
+    #[serde(flatten)]
+    pub action: RequestAction,
+
+    /// The ID of the device requesting the event.
+    pub requesting_device_id: DeviceIdBox,
+
+    /// A random string uniquely identifying (with respect to the requester and the target) the
+    /// target for a secret.
+    ///
+    /// If the secret is requested from multiple devices at the same time, the same ID may be used
+    /// for every target. The same ID is also used in order to cancel a previous request.
+    pub request_id: String,
+}
+
+impl RequestToDeviceEventContent {
+    /// Creates a new `RequestToDeviceEventContent` with the given action, requesting device ID and
+    /// request ID.
+    pub fn new(
+        action: RequestAction,
+        requesting_device_id: DeviceIdBox,
+        request_id: String,
+    ) -> Self {
+        Self { action, requesting_device_id, request_id }
+    }
+}
+
+/// Action for a *m.secret.request* event.
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Deserialize)]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+#[serde(try_from = "RequestActionInit")]
+pub enum RequestAction {
+    /// Request a secret by its name.
+    Request(SecretName),
+
+    /// Cancel a request for a secret.
+    RequestCancellation,
+
+    #[doc(hidden)]
+    _Custom(String),
+}
+
+impl Serialize for RequestAction {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let mut st = serializer.serialize_struct("request_action", 2)?;
+
+        match self {
+            Self::Request(name) => {
+                st.serialize_field("name", name)?;
+                st.serialize_field("action", "request")?;
+                st.end()
+            }
+            Self::RequestCancellation => {
+                st.serialize_field("action", "request_cancellation")?;
+                st.end()
+            }
+            RequestAction::_Custom(custom) => {
+                st.serialize_field("action", custom)?;
+                st.end()
+            }
+        }
+    }
+}
+
+#[derive(Deserialize)]
+struct RequestActionInit {
```

Comment:
I'd prefer a different name since our `Init` structs elsewhere are public types with `From`, not `TryFrom` being used to convert them to the "main" type.

```suggestion
struct RequestActionJsonRepr {
```